HollowEngine/BulletHell/src/DungeonGeneration/DungeonFloor.cpp

```diff
@@ -0,0 +1,344 @@
+#include "DungeonFloor.h"
+#include "Hollow/Utils/Random.h"
+#include "Hollow/Utils/UniqueID.h"
+
+namespace BulletHell
+{
+    DungeonFloor::DungeonFloor()
+        : mWidth(0)
+        , mHeight(0)
+        , mFloorNum(0)
+    {
+    }
+
+    DungeonFloor::DungeonFloor(int width, int height, int floorNum)
+        : mWidth(width)
+        , mHeight(height)
+        , mFloorNum(floorNum)
+    {
+        int size = width * height;
+        mRooms.reserve(size);
+        for (int i = 0; i < size; i++)
+        {
+            mRooms.emplace_back();
+        }
+    }
+
+    void DungeonFloor::Generate(int numRooms)
+    {
+        if (numRooms < 2)
+        {
+            std::cout << "Error! Dungeon Floor Generator: Less then 2 rooms" << std::endl;
+            return;
+        }
+        if (numRooms > mWidth * mHeight)
+        {
+            std::cout << "Error! Dungeon Floor Generator: To many rooms, to little dungeon floor size" << std::endl;
+            return;
+        }
+
+        ////// Room generation
+
+        //// entrance room
+        unsigned seed = 521288629;
+        Random::SetSeed(seed);
+        auto entranceRow = Random::RangeSeeded(0, mHeight - 1);
+        auto entranceCol = Random::RangeSeeded(0, mWidth - 1);
+        const int entranceIndex = Index(entranceRow, entranceCol);
+        mRooms[entranceIndex].Set(DungeonRoomType::ENTRANCE, Hollow::GenerateUniqueID<DungeonRoom>(), 0, mFloorNum, entranceRow, entranceCol);
+
+        // precompute distance of other rooms from entrance
+        for (int col = 0; col < mWidth; col++)
+        {
+            for (int row = 0; row < mHeight; row++)
+            {
+                mRooms[Index(row, col)].mDistFromEntrance = std::abs(row - entranceRow) + std::abs(col - entranceCol);
+            }
+        }
+
+        // generate a pool of rooms to pick from next
+        std::vector<int> possibleRoomLocation; // pool of rooms that could be added to the dungeon(indexis to 1D array of rooms)        
+        if (entranceCol > 0)
+        {
+            int roomCandidateIndex = Index(entranceRow, entranceCol - 1); // index of a room that might be added to the pool of room to pick from
```

Comment:
These 3 lines of code could be turned into a function. Something like AddRoomCandidate(int rowOffset, int colOffset).